src/main/scala/io/conduktor/ksm/source/HttpSourceAcl.scala

```diff
@@ -0,0 +1,216 @@
+    * Note: the first call to this function should never return `None`.
+    *
+    * Kafka Security Manager will not update ACLs in Kafka until there are no errors in the result
+    *
+    * @return
+    */
+  override def refresh(): Option[ParsingContext] = {
+
+    val request: Request = new Request(uri)
+    // super important in order to properly fail in case a timeout happens for example
+    request.enableThrowingIOException(true)
+    request.followRedirects(false)
+    request.connectTimeoutMillis(Int.MaxValue)
+    request.readTimeoutMillis(Int.MaxValue)
+
+    if (authenticationType == AuthenticationType.GOOGLE_IAM) {
+      request.header("Authorization", getIdToken())
+    }
+
+    // we use this header for the 304
+    lastModified.foreach(header => request.header("If-Modified-Since", header))
+    request.header("Content-Type", "text/plain") // only type supported for now
+    val response: Response = HTTP.request(httpMethod, request)
+
+    response.status match {
+      case 200 =>
+        lastModified = response.header("Last-Modified")
+
+        validateBodyLength(response)
+
+        Some(
+          ParsingContext(
+            parserRegistry.getParser(this.parser),
+            new BufferedReader(new StringReader(response.textBody))
+          )
+        )
+      case 304 =>
+        None
+      case _ =>
+        // we got an http error so we propagate it
+        log.error("HTTP Status {} in HTTP Source request: {}", response.status, response.asString)
+        throw HTTPException(Some(response.asString), response)
+    }
+  }
+
+  private def validateBodyLength(response: Response): Unit = {
+    val optContentLengthHeader = response.headers
+      .find(h => CONTENT_LENGTH.equalsIgnoreCase(h._1))
+      .map(h => h._2)
+      .map(l => l.toInt)
+    if (optContentLengthHeader.isEmpty) {
+      log.warn(s"Response doesn't contain $CONTENT_LENGTH header, only contained the following headers: ${response.headers.keySet}. Skipping validation...")
+      return
+    }
+
+    val contentLengthHeader = optContentLengthHeader.get
+    val bodyLength = response.asBytes.length
+    log.info(s"Validating body length ($bodyLength bytes) received from $uri against Content-Length header ($contentLengthHeader bytes) claimed in response")
+    try {
+      val reader = new BufferedReader(new StringReader(response.textBody))
+      val aclLineCount = reader.lines.count
+      log.info("There were {} lines in the response received from {}", aclLineCount, uri)
+      reader.close()
+    }
+    catch {
+      case e: Exception => log.warn(s"Failed to compute number of lines in the response received from $uri", e)
+    }
+    if (contentLengthHeader != bodyLength) {
+      val errorMessage = s"Body length ($bodyLength bytes) inconsistent with Content-Length header ($contentLengthHeader bytes)"
+      log.error(errorMessage)
+      throw HTTPException(Some(errorMessage), response)
+    }
+  }
+
+  private def getIdToken(): String = {
```

Comment:
we want to make addind new authentication easy. What do you think of isolating the authentication code behind an interface:
```scala
trait HttpAuthentication {
  def authHeader: String
}

class GoogleIAM(serviceAccountName: String, targetAudience: String, serviceAccountKey: String) extends HttpAuthentication {
```

and in `HttpSource` class:
```scala
  var authentication: Option[HttpAuthentication]

   authentication = config.getString(AUTHENTICATION_TYPE) match {
      case AuthenticationType.GOOGLE_IAM => Some(new GoogleIAM("TODO", "TODO", "TODO")) 
      case AuthenticationType.NONE => None
   }

```
(There should be no settings specific to google directly in the HttpSource class)

Reply:
Good point. Let me know if the implementation is ok

Reply:
looks ok :+1: